### PR TITLE
add groups tokenfile

### PR DESCRIPTION
**Pull request by gambol99**
Summary: allowing for groups to be added to the user tokefile. So it can be later be consumed from ABAC?

[plugin/auth/tokenfile]
- allowing for variable length groups to be added to the static token file  

[docs/admin/authentication]
- updating the documentation for token file

### CODE REVIEW

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go [hunk 1]
```diff
@@ -66,6 +73,18 @@ token2,user2,uid2
 		} else if !reflect.DeepEqual(testCase.User, user) {
 			t.Errorf("%d: expected user %#v, got %#v", i, testCase.User, user)
 		}
+
+		if testCase.User.Groups != nil {
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: `sets.NewString(testCase.User.Groups...).Equals(sets.NewString(user.GetGroups()...)`?

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile.go [hunk 2]
```diff
@@ -40,6 +40,8 @@ func NewCSV(path string) (*TokenAuthenticator, error) {
 
 	tokens := make(map[string]*user.DefaultInfo)
 	reader := csv.NewReader(file)
+	reader.FieldsPerRecord = -1
+	reader.Comment = '#'
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile.go]: not sure we want to enable comments in these files... this doesn't seem related to this PR

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go (hunk 1 above)
Comment (gambol99) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: That's a lot better ... although, looking at it again i can probably drop the test altogether and rely on the original to catch it

```go
} else if !reflect.DeepEqual(testCase.User, user) {
    t.Errorf("%d: expected user %#v, got %#v", i, testCase.User, user)
}
```

**Comment by wojtek-t** [comment]
@k8s-bot ok to test

File: docs/admin/authentication.md [hunk 3]
```diff
@@ -46,7 +46,8 @@ to apiserver.  Currently, tokens last indefinitely, and the token list cannot
 be changed without restarting apiserver.
 
 The token file format is implemented in `plugin/pkg/auth/authenticator/token/tokenfile/...`
-and is a csv file with 3 columns: token, user name, user uid.
+and is a csv file with a minimum of 3 columns: token, user name, user uid, followed by an optional
+variable number of groups
```
Comment (vaijab) [inline docs/admin/authentication.md]: A full stop is missing :-)

**Comment by liggitt** [comment]
lgtm, squash to a single commit

**Comment by gambol99** [comment]
I've squashed into a single commit

**Comment by liggitt** [comment]
```
/home/travis/gopath/src/github.com/kubernetes/kubernetes/docs/ is out of date. Please run hack/update-generated-docs.sh

!!! Error in hack/verify-generated-docs.sh:28
```

**Comment by liggitt** [comment]
update generated docs and make sure verify tests run cleanly

**Comment by gambol99** [comment]
Removed the whitespace at the end of line which was the causing issue ... tests run fine now 

```shell
[jest@starfury kubernetes]$ hack/verify-generated-docs.sh 
+++ [1015 19:20:23] Building go targets for linux/amd64:
    cmd/gendocs
    cmd/genman
    cmd/genbashcomp
    cmd/mungedocs
+++ [1015 19:20:23] Placing binaries
Generated docs are up to date.
```

**Comment by wojtek-t** [comment]
@gambol99 - can you please rebase your PR - this should solve the Shippable failures (it helped in few other cases)

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go [hunk 4]
```diff
@@ -52,9 +54,13 @@ token2,user2,uid2
 		},
 		{
 			Token: "token3",
+			User:  &user.DefaultInfo{Name: "user3", UID: "uid3", Groups: []string{"group1", "group2"}},
+			Ok:    true,
 		},
 		{
 			Token: "token4",
+			User:  &user.DefaultInfo{Name: "user4", UID: "uid4", Groups: []string{"group2"}},
+			Ok:    true,
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: need a test of an invalid token (all your test cases are valid tokens now)

Comment (gambol99) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: I've copied in the old false tests from master and squashed down to a single commit again

**Comment by wojtek-t** [comment]
Jenkins failure is a flake:
"Job should run a job to completion when tasks sometimes fail and are locally restarted"

@k8s-bot test this please

**Comment by wojtek-t** [comment]
Jenkins failure is an unrelated flake:
"Pods should have their auto-restart back-off timer reset on image update"

@k8s-bot test this please

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go [hunk 5]
```diff
@@ -29,6 +29,8 @@ func TestTokenFile(t *testing.T) {
 	auth, err := newWithContents(t, `
 token1,user1,uid1
 token2,user2,uid2
+token3,user3,uid3,group1,group2
```
Comment (erictune) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: We can never add another column to this now, since group1 is repeated.

Comment (erictune) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: if you used double quotes around a comma-separated list of groups, then this would not be a problem.

Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: strike now while the iron is hot

**Comment by erictune** [comment]
@gambol99 can you send a PR soon which modifies the format so that it does not break extensibility.

File: plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go (hunk 5 above)
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/tokenfile/tokenfile_test.go]: lists inside lists is tricky, but I agree about not locking ourselves in

**Comment by gambol99** [comment]
Will do. Though I did raise the PR #15972 as an alternative / replacement .. since json|yaml is easier to deal with
